server/impl/src/main/java/com/walmartlabs/concord/server/security/ldap/ConcordLdapContextFactory.java

```diff
@@ -23,26 +23,118 @@
+import javax.naming.directory.Attribute;
+import javax.naming.directory.Attributes;
+import javax.naming.directory.DirContext;
+import javax.naming.directory.InitialDirContext;
 import javax.naming.ldap.LdapContext;
 import java.util.Hashtable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import static com.walmartlabs.concord.server.security.ldap.Utils.recursionLimiter;
 
 @Named
 public class ConcordLdapContextFactory implements LdapContextFactory {
 
-    private final LdapContextFactory delegate;
+    private LdapContextFactory delegate;
+
+    private static final Logger log = LoggerFactory.getLogger(ConcordLdapContextFactory.class);
+
+    private static final String protocol = "ldaps";
+    private static final String port = "3269";
+
+    private Iterator<String> ldapUrlIterator;
+
+    private final LdapConfiguration cfg;
 
+    /*
+    1. check dnsSRV and refresh SRV list
+    2. pic one and create LDAP context
+    3. while getContext check for communication exception and update context with next URL in SRV list
+    4. on SRV list empty, call step 1.
+     */
     @Inject
     @SuppressWarnings("unchecked")
-    public ConcordLdapContextFactory(LdapConfiguration cfg) {
+    public ConcordLdapContextFactory(LdapConfiguration cfg) throws NamingException {
+        this.cfg = cfg;
+
+        this.refreshSRVList();
+
+        this.setLdapContextFactory(getNewContextInstance(resolveUrl()));
+    }
+
+    public String getCurrentLdapUrl() {
+        return ((JndiLdapContextFactory) this.delegate).getUrl();
+    }
+
+    public synchronized void setLdapContextFactory(LdapContextFactory ldapContextFactory) {
+        this.delegate = ldapContextFactory;
+    }
+    
+    public synchronized void setLdapUrlIterator(Iterator<String> ldapUrlIterator) {
+        this.ldapUrlIterator = ldapUrlIterator;
+    }
+    
+    @Override
+    public LdapContext getSystemLdapContext() throws NamingException {
+        try {
+            return this.delegate.getSystemLdapContext();
+        } catch (CommunicationException e) {
+            recursionLimiter();  // limiter to safe guard from infinite stack overflow
+            handleCommunicationException(e);
+            return getSystemLdapContext();
+        }
+    }
+
+    @Override
+    public LdapContext getLdapContext(String username, String password) throws NamingException {
+        try {
+            return this.delegate.getLdapContext(username, password);
+        } catch (CommunicationException e) {
+            recursionLimiter();
+            handleCommunicationException(e);
+            return getLdapContext(username, password);
+        }
+    }
+
+    @Override
+    public LdapContext getLdapContext(Object principal, Object credentials) throws NamingException {
+        try {
+            return this.delegate.getLdapContext(principal, credentials);
+        } catch (CommunicationException e) {
+            recursionLimiter();
```

Comment:
This is concerning. In which situations this call might be recursive?

Reply:
no good servers(may be decommissioned/removed off) in the list then call gets recursive to check for each server

Reply:
Why it cannot be a loop? Sifting through stacktraces seems a bit unconventional.

Reply:
okay. was trying to reduce code duplication. [Replaced](https://github.com/walmartlabs/concord/pull/412/commits/916fae77478346b7620149bce8f318b39c53e014) with loop approach to achieve depth limit